[AArch64] Move BSL generation to lowering.

davemgreen opened the pull request:
It is generally better to allow the target independent combines before creating AArch64 specific nodes (providing they don't mess it up). This moves the generation of BSL nodes to lowering, not a combine, so that intermediate nodes are more likely to be optimized. There is a small change in the constant handling to detect legalized buildvector arguments correctly.

Fixes #149380 but not directly. #151856 contains a direct fix for expanding the pseudos.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -14737,12 +14737,109 @@ static SDValue tryLowerToSLI(SDNode *N, SelectionDAG &DAG) {
+
+      if (ISD::isConstantSplatVector(N0->getOperand(i).getNode(), Val1) &&
+          ISD::isConstantSplatVector(N1->getOperand(j).getNode(), Val2) &&
+          ~Val1.trunc(Bits) == Val2.trunc(Bits)) {
```
I see this function is not entirely a case of moving. Can you pre-commit the moving of this function as a non-functional change, and rebase this one on top?

rj-jesus reviewed (commented):
Hi @davemgreen, I think this sounds like a good idea. The free variable cases I looked at recently should be unaffected by this change in any case. :)

davemgreen committed:
[AArch64] Move BSL generation to lowering.

It is generally better to allow the target independant combines before lowering
to AArch64 specific ones (providing they don't mess it up). This moves the
generation of BSL nodes to lowering, not a combine, so that intermediate nodes
are more likely to be optimized. There is a small change in the constant
handling to detect legalized buildvector arguements correctly.

davemgreen commented:
I've rebased over 6f272d1ecf70fc555efb1a0c601095031d5b2ca9 and addressed some comments. Thanks.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -14742,23 +14742,15 @@ static SDValue tryLowerToSLI(SDNode *N, SelectionDAG &DAG) {
   if (VT.isScalableVector() && !Subtarget.hasSVE2())
     return SDValue();
 
-  if (VT.isFixedLengthVector() &&
```
Did you remove this condition on purpose?

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -14808,14 +14800,13 @@ static SDValue tryCombineToBSL(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
       if (ISD::isConstantSplatVector(N0->getOperand(i).getNode(), Val1) &&
           ISD::isConstantSplatVector(N1->getOperand(j).getNode(), Val2) &&
-          (BitMask & ~Val1.getZExtValue()) == Val2.getZExtValue()) {
+          ~Val1.trunc(Bits) == Val2.trunc(Bits)) {
```
I guess the issue here and below was that the mask wasn't applied to both sides? Either way, I think the new version looks good.

rj-jesus reviewed (approved):
As far as I can tell this looks good to me, thanks!
